Ray Surface FE Intersection

gzagaris opened the pull request:
This PR adds a routine for computing the intersection of a given Ray with a surface FE element. The approach employs a Newton-Raphson iterative solve with backtracking line search for increased robustness. The backtracking linesearch uses the strong Wolfe condition criteria for determining an appropriate stepsize.

tzanio commented on linalg/densemat.cpp:
```diff
@@ -3971,7 +4053,12 @@ void LUFactors::Factor(int m)
       }
-      MFEM_ASSERT(data[i+i*m] != 0.0, "division by zero");
+
+      if ( abs(data[i+i*m]-0.0) <= 1.e-9 )
```
Not sure about the implications of this magic number

gzagaris replied:
This was done instead of `if ( data[i+i*m] == 0.0 )`. What would you suggest? 

I agree magic numbers are generally bad. We could have an optional argument to the `LUFactors::Factor(int m, double TOL=1.e-9)` that way the application can set it. Is this preferable?

gzagaris replied:
I added an optional tolerance argument to the `LUFactors::Factor` method. This avoids the magical numerical value in this check. The default is still `1.e-9`, but the caller may change this. Let me know what you think.

tzanio commented on fem/fe_ray_intersect.hpp:
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC. Produced at
```
Does this need to be in a separate file? Shouldn't it be a part of the `ElementTransfromation` routines? Will the algorithm sill work if the element transformation is not given by isoparametric finite elements (but we have its Jacobian, etc.)?

gzagaris replied:
Thanks @tzanio , I did take a look at the `ElementTransformation` routines, but, I wasn't sure if this type of operation would fit with the design of the methods therein. So, I added this in a separate file primarily because that seemed like the simplest approach and point of entry for me.

In this initial design, an application may use this as follows:
```
// set any parameters (optional, reasonable defaults are provided)
mfem::fe_ray_setMaxNewtonIteration( 20 );
// etc.
...
const FiniteElement* fe = mesh->GetNodalFESpace()->GetFE( ielem );
bool converged = mfem::fe_ray_solve( ielem, mesh, x0, n, r, t );
if ( converged && mfem::fe_ray_intersects( fe, r, t ) )
{
   // compute intersection point (assuming 3D)
   p[0]=x0[0]+n[0]*t;
   p[1]=x0[1]+n[1]*t;
   p[2]=x0[2]+n[2]*t;

    ....
}

...

```

How would you suggest to change the user-facing API if this becomes part of the `ElementTransformation` routines? 

From a style perspective, these routines do not hold any state information, so, it feels a bit clunky to require allocating an object to call a function instead of calling a function. Perhaps, they can be *static* methods of a class, that way they can be encapsulated within a namespace accordingly ?

The impetus for keeping a C-Style API more or less is motivated by the need to call these routines from within a kernel loop (projecting in to the future). We would typically have a bunch of rays whose intersection with a surface needs to be computed. Under this scenario, each thread would operate on a single ray checking a list of candidate elements for the intersection by calling `fe_ray_solve` and `fe_ray_intersects` respectively.

You make a good point about whether isoparametric elements are required or not. Given that we can evaluate a jacobian, as you mentioned, I would expect that this should work.  I'll add some tests using a different basis to verify.

gzagaris replied:
@tzanio I added a tests using `H1Pos_FiniteElementCollection` basis, in addition to the `H1_FECollection`, i.e., `Lagrange` basis that I was using before. However, I think those are still considered iso-parametric finite elements? Is that correct? 

Could you clarify what you meant by "element transformation not given by iso-parametric finite elements"? Is there a different `FiniteElementCollection` that we should have a test for?

tzanio commented on fem/fe_ray_intersect.cpp:
```diff
@@ -0,0 +1,453 @@
+//------------------------------------------------------------------------------
+bool check_convergence( const double* dx, int ndims, double TOL )
+{
+   double l1norm = 0.0;
```
Lots of free functions in this file. Is there a natural way to put them in a class, e.g. something like `InverseElementTransformation`?

gzagaris replied:
Most of them are private to this implementation and are encapsulated within an anonymous namespace. The methods that are exposed to the application are defined in the ``*.hpp`` file.

They can be put in the class, but, I think from the user's perspective, it would be simpler if they were static methods of the class (see my previous response which relates to this).

tzanio commented:
@gzagaris -- if this is still relevant, do you want to update it and mark as `ready-for-review`?

gzagaris commented:
> @gzagaris -- do you want to update this and mark as `ready-for-review`?

Thanks @tzanio, I rebased this branch to the latest and updated the test. 

At your convenience, I would like to get some feedback on the general API/usability of the functionality introduced in this PR. If it is easier, we can also discuss more about this in person. I'll be back in the office on Monday 1/6.

Also, how do I mark this as `ready-for-review`? Is that a `Label`? It looks like I cannot edit the labels...

Thanks again for your help.

gzagaris commented:
It looks like `ex18p` is failing on a Mac when `DEBUG=NO MPI=YES`. This should be orthogonal to the changes herein. The unit tests are passing. Any ideas?

gzagaris commented:
> It looks like `ex18p` is failing on a Mac when `DEBUG=NO MPI=YES`. This should be orthogonal to the changes herein. The unit tests are passing. Any ideas?

All tests passed after triggering the tests a 2nd time.

tzanio commented:
> It looks like `ex18p` is failing on a Mac when `DEBUG=NO MPI=YES`. This should be orthogonal to the changes herein. The unit tests are passing. Any ideas?

Seems fine in Travis, and on my MacBook... Can you provide more details?

tzanio commented:
> Also, how do I mark this as ready-for-review? Is that a Label? It looks like I cannot edit the labels...

Yes, this is an existing label. You should be able to add it yourself when ready...

gzagaris commented:
> Seems fine in Travis, and on my MacBook... Can you provide more details?

Ditto -- it seems like a random failure, since things passed the 2nd time. Is this test known to have sporadic failures?

gzagaris commented:
> Yes, this is an existing label. You should be able to add it yourself when ready...

Thanks, that's what I thought. Unfortunately,  I cannot edit and add new labels.

tzanio commented:
Lets discuss in person when you get back.

gzagaris commented:
@tzanio , @v-dobrev , I am going to split this PR into two. I am moving some of the minor linear algebra changes, that are somewhat orthogonal to this, in PR #1243 to make it easier to review.

gzagaris commented on fem/intrules.hpp:
```diff
@@ -28,6 +28,22 @@ class IntegrationPoint
 
+   IntegrationPoint() = default;
+
+   IntegrationPoint (const double* p, int dim)
```
I added custom constructors to create an `IntegrationPoint` from a `double*` and  a given dimension, `dim` and ` double*` and weight.  This was prior to the  introduction of the `index` member variable. 

I am assuming the `index` is according to the local numbering scheme of an element. Is this assumption correct?

How should we initialize this here? 
* Just set it to zero?
* Make it an optional argument to the new constructors, which defaults to zero if not specified?

For the particular use-case, I am not making use of the "index".

v-dobrev replied:
Yes, `index` can be set in the context of an `IntegrationRule` according to a local numbering.

It is probably best to use `-1` for `index` to indicate invalid/undefined value, since `0` is a valid value.

Regarding the addition of an optional parameter in the constructor for the `index` -- either choice is fine -- use what you prefer.

gzagaris replied:
Sounds good --  thanks @v-dobrev
